File: python/mirage/core/bin/read.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.accessor.bin import BinAccessor
from mirage.cache.index import NULL_INDEX, IndexCacheStore
from mirage.core.bin.render import render_stub
from mirage.types import PathSpec
from mirage.utils.errors import eisdir, enoent


async def read(accessor: BinAccessor,
               path: PathSpec,
               index: IndexCacheStore = NULL_INDEX) -> bytes:
    """Render one program's file.

    Args:
        accessor (BinAccessor): Accessor holding the lookup.
        path (PathSpec): Virtual path under the view.
        index (IndexCacheStore): Unused; op signature parity.

    Returns:
        bytes: The program's stub, fresh on every call.
    """
    key = path.mount_path.strip("/")
    if key == "":
        raise eisdir(path)
    note = None if "/" in key else accessor.note(key)
    if note is None:
        raise enoent(path)
    return render_stub(key, note)

File: typescript/packages/core/src/core/bin/read.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { BinAccessor } from '../../accessor/bin.ts'
import type { PathSpec } from '../../types.ts'
import { eisdir, enoent } from '../../utils/errors.ts'
import { stripSlash } from '../../utils/slash.ts'
import { renderStub } from './render.ts'

/** Render one program's file, fresh on every call. */
export function read(accessor: BinAccessor, path: PathSpec): Promise<Uint8Array> {
  const key = stripSlash(path.mountPath)
  if (key === '') return Promise.reject(eisdir(path))
  const note = key.includes('/') ? null : accessor.note(key)
  if (note === null) return Promise.reject(enoent(path))
  return Promise.resolve(renderStub(key, note))
}
